### PR TITLE
Refactor Generic Host lifetimes to work better with ServiceBase 

### PR DESCRIPTION
**Pull request by Tratcher**
#1347
The RegisterDelayStartCallbackAsync callback is inlined to work around the ServiceBase.Run blocking call. 
RegisterStopCallback goes away, it only ever called StopApplication.
ProcessLifetime goes away, ConsoleLifetime becomes the default.

### CODE REVIEW

**Comment by Tratcher** [comment]
@glennc one funny thing I noticed when debugging is that OnStop gets called 3 times, but it's fine because the CTS in StopApplication prevents any recursion.
1. OnStop called by the service.
2. OnStop => StopApplication => Stop => OnStop
3. Host.StopAsync => Stop => OnStop

**Comment by davidfowl** [comment]
StartAsync cannot block. The usage pattern is too different from the other usage pattern. We need to pick one way and stick with it.

**Comment by Tratcher** [comment]
Easy enough, I forced ServiceBase.Run onto a background thread.

File: samples/GenericHostSample/ServiceBaseLifetime.cs [hunk 1]
```diff
@@ -23,23 +23,25 @@ public static Task RunAsServiceAsync(this IHostBuilder hostBuilder, Cancellation
...
 
-            Run(this);
-        }
+            ApplicationLifetime.ApplicationStopping.Register(Stop);
 
-        public void RegisterStopCallback(Action<object> callback, object state)
-        {
-            _stopCallback = callback ?? throw new ArgumentNullException(nameof(callback));
-            _stopState = state;
+            Task.Run(() => Run(this)); // Otherwise this would block and prevent Start from finishing.
```
Comment (davidfowl) [inline samples/GenericHostSample/ServiceBaseLifetime.cs]: Spawn a real thread so we don't block a thread pool thread.

Comment (Tratcher) [reply samples/GenericHostSample/ServiceBaseLifetime.cs]: I uncovered a design flaw when testing this. RegisterDelayStartCallbackAsync has no abort mechanic. If you try to run your service app outside of a service Run(this) pops up a dialog box to tell you what you did wrong, but Run itself doesn't throw, it exits early. OnStart is never called so the delay start callback is never called, and the app hangs in StartAsync.

**Comment by Tratcher** [comment]
Redesigned and Updated. I got rid of all the callbacks.

File: test/Microsoft.Extensions.Hosting.Tests/HostTests.cs [hunk 2]
```diff
@@ -414,7 +353,6 @@ public void HostCanBeStarted()
 
                 lifetime = (FakeHostLifetime)host.Services.GetRequiredService<IHostLifetime>();
                 Assert.Equal(1, lifetime.StartCount);
-                Assert.Equal(1, lifetime.StoppingCount);
```
Comment (davidfowl) [inline test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: Why did this change?

Comment (Tratcher) [reply test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: Stopping was one of the test callbacks we removed.

Comment (davidfowl) [reply test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: What do you mean? The test itself changed?

Comment (Tratcher) [reply test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: I merged RegisterStopCallback and RegisterDelayStartCallback into the new WaitForStartAsync so there's one less event to mock in the tests.

Comment (davidfowl) [reply test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: Oh, I thought this was IApplicationLifetime

**Pull request merged**